localstack/services/s3/presigned_url.py

```diff
@@ -502,60 +351,82 @@ def validate_presigned_url_s3(context: RequestContext) -> None:
+    :return: headers in HTTPHeaders format
     """
-    if headers.get("x-amz-content-sha256", None) == "STREAMING-AWS4-HMAC-SHA256-PAYLOAD":
-        # this is sign that this is a SigV4 request, with payload encoded
-        # we do not support payload encoding yet
-        # moto parses it to an int, it would raise a 500
-        content_length = headers.get("x-amz-decoded-content-length")
-        if not content_length:
-            raise SignatureDoesNotMatch('"X-Amz-Decoded-Content-Length" header is missing')
-        try:
-            int(content_length)
-        except ValueError:
-            raise SignatureDoesNotMatch('Wrong "X-Amz-Decoded-Content-Length" header')
+    # Werkzeug Headers can have multiple values for the same key
+    # HTTPHeaders will append automatically the values when we set it to the same key multiple times
+    # see https://docs.python.org/3/library/http.client.html#httpmessage-objects
+    # see https://docs.python.org/3/library/email.compat32-message.html#email.message.Message.__setitem__
+    headers = HTTPHeaders()
+    for key, value in werkzeug_headers.items():
+        headers[key] = value
+
+    return headers
 
 
-def _get_signature_of_presigned_request_s3v4(
-    context: RequestContext, request_netloc: str
-) -> FindSigV4Result:
+def _reverse_inject_signature_hmac_v1_query(context: RequestContext) -> Request:
     """
-    Returns the signature of the request
-    :param context: RequestContext
-    :param request_netloc: the host of the original request
+    Reverses what does HmacV1QueryAuth._inject_signature while injecting the signature in the request.
+    Transforms the query string parameters in headers to recalculate the signature
+    see botocore.auth.HmacV1QueryAuth._inject_signature
+    :param context:
     :return:
     """
-    # if both x-amz* header and query param, InvalidRequest, conflicting -> test it
-    query_parameters = context.request.args
 
-    credentials = ReadOnlyCredentials(
-        TEST_AWS_ACCESS_KEY_ID,
-        TEST_AWS_SECRET_ACCESS_KEY,
-        query_parameters.get("X-Amz-Security-Token", None),
+    new_headers = {}
+    new_query_string_dict = {}
+
+    for header, value in context.request.args.items():
+        header_low = header.lower()
+        if header_low not in HmacV1QueryAuthValidation.post_signature_headers:
+            new_headers[header] = value
+        elif header_low in HmacV1QueryAuthValidation.QSAOfInterest_low:
+            new_query_string_dict[header] = value
+
+    # there should not be any headers here. If there are, it means they have been added by the client
+    # We should verify them, they will fail the signature except if they were part of the original request
+    for header, value in context.request.headers.items():
+        header_low = header.lower()
+        if header_low.startswith("x-amz-") or header_low in ["content-type", "date", "content-md5"]:
+            new_headers[header_low] = value
```

Comment:
not sure what we are doing with the resulting request, but should we maybe preserve the original header casing?
```suggestion
            new_headers[header] = value
```

Reply:
While creating the signature, it lowers the name of the header as well, so we need to lower the header to be able to match the signature. The corresponding code snippet is here in botocore:
`botocore.auth.SigV4Auth.headers_to_sign` (l.231)
```
def headers_to_sign(self, request):
    """
    Select the headers from the request that need to be included
    in the StringToSign.
    """
    header_map = HTTPHeaders()
    for name, value in request.headers.items():
        lname = name.lower()
        if lname not in SIGNED_HEADERS_BLACKLIST:
            header_map[lname] = value
    if 'host' not in header_map:
        # TODO: We should set the host ourselves, instead of relying on our
        # HTTP client to set it for us.
        header_map['host'] = _host_from_url(request.url)
    return header_map
```
I could add a comment pointing to this. The code is full of traps to reverse. I had to use the lower-cased one or it would fail in CI if I remember well.